File: packages/core-java/src/main/java/io/github/mertc0604/converttools/coordinates/Coordinates.java
package io.github.mertc0604.converttools.coordinates;

import io.github.mertc0604.converttools.coordinates.core.CoordinateMath;
import io.github.mertc0604.converttools.coordinates.format.CoordinateAxis;
import io.github.mertc0604.converttools.coordinates.format.DmsFormat;
import io.github.mertc0604.converttools.coordinates.grid.Gars;
import io.github.mertc0604.converttools.coordinates.grid.Georef;
import io.github.mertc0604.converttools.coordinates.grid.Mgrs;
import io.github.mertc0604.converttools.coordinates.model.AngularCell;
import io.github.mertc0604.converttools.coordinates.model.CoordinateFormat;
import io.github.mertc0604.converttools.coordinates.model.CoordinatePoint;
import io.github.mertc0604.converttools.coordinates.model.CoordinateResolution;
import io.github.mertc0604.converttools.coordinates.model.CoordinateResult;
import io.github.mertc0604.converttools.coordinates.model.CoordinateSource;
import io.github.mertc0604.converttools.coordinates.model.GridCell;
import io.github.mertc0604.converttools.coordinates.model.GridCoordinate;
import io.github.mertc0604.converttools.coordinates.projection.UtmUps;
import io.github.mertc0604.converttools.geodesy.GeoPoint;

import java.util.EnumMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;

public final class Coordinates {
    private static final int DD_DIGITS = 10;
    private static final int DMS_SECOND_DIGITS = 5;
    private static final int DDM_MINUTE_DIGITS = 7;
    private static final int GRID_METRE_DIGITS = 3;
    private static final double GARS_CELL_DEGREES = 1.0 / 12;
    private static final int GEOREF_PRECISION = 4;
    private static final Pattern ZONE = Pattern.compile("^\\d{1,2}$");

    private Coordinates() {
    }

    public static CoordinateSource fromDecimalDegrees(
            String latitude,
            String longitude
    ) {
        return CoordinateSource.point(
                new CoordinatePoint(
                        CoordinateMath.parseNumber(latitude, "Latitude"),
                        CoordinateMath.parseNumber(longitude, "Longitude")
                )
        );
    }

    public static CoordinateSource fromDecimalDegrees(
            double latitude,
            double longitude
    ) {
        return CoordinateSource.point(
                new CoordinatePoint(latitude, longitude)
        );
    }

    public static CoordinateSource fromDms(
            String latitude,
            String longitude
    ) {
        return CoordinateSource.point(
                DmsFormat.parsePair(latitude, longitude)
        );
    }

    public static CoordinateSource fromDdm(
            String latitude,
            String longitude
    ) {
        return CoordinateSource.point(
                DmsFormat.parsePair(latitude, longitude)
        );
    }

    public static CoordinateSource fromMgrs(String value) {
        GridCell decoded = Mgrs.decode(value, true);
        return CoordinateSource.gridCell(
                decoded.point(),
                decoded.cellMetres()
        );
    }

    public static CoordinateSource fromUtmUps(
            String zoneValue,
            String hemisphereValue,
            String eastingValue,
            String northingValue
    ) {
        String source = zoneValue.trim();
        if (!ZONE.matcher(source).matches()) {
            throw new IllegalArgumentException(
                    "Zone must be an integer from 0 to 60."
            );
        }
        int zone = Integer.parseInt(source);
        if (zone < 0 || zone > 60) {
            throw new IllegalArgumentException(
                    "Zone must be an integer from 0 to 60."
            );
        }
        String hemisphere = hemisphereValue.trim()
                .toUpperCase(Locale.ROOT);
        if (!"N".equals(hemisphere) && !"S".equals(hemisphere)) {
            throw new IllegalArgumentException(
                    "Hemisphere must be N or S."
            );
        }
        return CoordinateSource.point(
                UtmUps.inverse(
                        zone,
                        "N".equals(hemisphere),
                        CoordinateMath.parseNumber(eastingValue, "Easting"),
                        CoordinateMath.parseNumber(northingValue, "Northing")
                )
        );
    }

    public static CoordinateSource fromGars(String value) {
        AngularCell decoded = Gars.decode(value);
        return CoordinateSource.area(
                decoded.center(),
                decoded.cellDegrees()
        );
    }

    public static CoordinateSource fromGeoref(String value) {
        AngularCell decoded = Georef.decode(value);
        return CoordinateSource.angularCell(
                decoded.center(),
                decoded.cellDegrees()
        );
    }

    public static GeoPoint toGeoPoint(CoordinateSource source) {
        Objects.requireNonNull(source, "source");
        return toGeoPoint(source.point());
    }

    public static GeoPoint toGeoPoint(CoordinatePoint point) {
        Objects.requireNonNull(point, "point");
        return new GeoPoint(point.latitude(), point.longitude());
    }

    public static CoordinatePoint fromGeoPoint(GeoPoint point) {
        Objects.requireNonNull(point, "point");
        return new CoordinatePoint(point.latitude(), point.longitude());
    }

    public static CoordinateResult results(
            CoordinatePoint point,
            int mgrsPrecision
    ) {
        return results(CoordinateSource.point(point), mgrsPrecision);
    }

    public static CoordinateResult results(
            CoordinateSource source,
            int mgrsPrecision
    ) {
        validateMgrsPrecision(mgrsPrecision);
        CoordinatePoint point = source.point();
        GridCoordinate grid = UtmUps.forward(
                point.latitude(),
                point.longitude()
        );
        String gridPrefix = grid.zone() == 0
                ? "UPS " + (grid.north() ? "N" : "S")
                : grid.zone() + (grid.north() ? "N" : "S");

        double ddStepDegrees = Math.pow(10, -DD_DIGITS);
        double dmsStepDegrees =
                Math.pow(10, -DMS_SECOND_DIGITS) / 3600;
        double ddmStepDegrees =
                Math.pow(10, -DDM_MINUTE_DIGITS) / 60;
        double mgrsCellMetres = Math.pow(10, 5 - mgrsPrecision);
        double gridStepMetres = Math.pow(10, -GRID_METRE_DIGITS);
        double georefCellDegrees =
                Math.pow(10, 2 - GEOREF_PRECISION) / 60;

        Map<CoordinateFormat, CoordinateResolution> resolution =
                new EnumMap<>(CoordinateFormat.class);
        resolution.put(
                CoordinateFormat.DD,
                CoordinateResolution.angularRounding(ddStepDegrees)
        );
        resolution.put(
                CoordinateFormat.DMS,
                CoordinateResolution.angularRounding(dmsStepDegrees)
        );
        resolution.put(
                CoordinateFormat.DDM,
                CoordinateResolution.angularRounding(ddmStepDegrees)
        );
        resolution.put(
                CoordinateFormat.MGRS,
                CoordinateResolution.gridCell(mgrsCellMetres)
        );
        resolution.put(
                CoordinateFormat.UTM_UPS,
                CoordinateResolution.gridRounding(gridStepMetres)
        );
        resolution.put(
                CoordinateFormat.GARS,
                CoordinateResolution.angularCell(GARS_CELL_DEGREES)
        );
        resolution.put(
                CoordinateFormat.GEOREF,
                CoordinateResolution.angularCell(georefCellDegrees)
        );

        return new CoordinateResult(
                point,
                CoordinateMath.fixed(point.latitude(), DD_DIGITS)
                        + ", "
                        + CoordinateMath.fixed(
                                point.longitude(),
                                DD_DIGITS
                        ),
                DmsFormat.formatDms(
                        point.latitude(),
                        CoordinateAxis.LATITUDE,
                        DMS_SECOND_DIGITS
                )
                        + "  "
                        + DmsFormat.formatDms(
                                point.longitude(),
                                CoordinateAxis.LONGITUDE,
                                DMS_SECOND_DIGITS
                        ),
                DmsFormat.formatDdm(
                        point.latitude(),
                        CoordinateAxis.LATITUDE,
                        DDM_MINUTE_DIGITS
                )
                        + "  "
                        + DmsFormat.formatDdm(
                                point.longitude(),
                                CoordinateAxis.LONGITUDE,
                                DDM_MINUTE_DIGITS
                        ),
                Mgrs.format(Mgrs.encode(point, mgrsPrecision)),
                gridPrefix
                        + "  "
                        + CoordinateMath.fixed(
                                grid.easting(),
                                GRID_METRE_DIGITS
                        )
                        + " E  "
                        + CoordinateMath.fixed(
                                grid.northing(),
                                GRID_METRE_DIGITS
                        )
                        + " N",
                Gars.encode(point),
                Georef.encode(point, GEOREF_PRECISION),
                resolution,
                source.kind(),
                source.cellMetres(),
                source.cellDegrees()
        );
    }

    private static void validateMgrsPrecision(int precision) {
        if (precision < 0 || precision > 5) {
            throw new IllegalArgumentException(
                    "MGRS precision must be between 0 and 5."
            );
        }
    }
}

File: packages/core-js/src/geodesy/coordinates.js
import { transformCrs } from "./projections/crs.js";
import {
  formatDdm,
  formatDms,
  parseCoordinatePair,
} from "./formats/dms.js";
import {
  decodeGars,
  encodeGars,
} from "./grids/gars.js";
import {
  decodeGeoref,
  encodeGeoref,
} from "./grids/georef.js";
import { decodeMgrs, encodeMgrs, formatMgrs } from "./grids/mgrs.js";
import {
  fixed,
  parseCoordinateNumber,
  pointFromValues,
} from "./core/numbers.js";
import {
  utmUpsForward,
  utmUpsInverse,
} from "./projections/utm-ups.js";

const DD_DIGITS = 10;
const DMS_SECOND_DIGITS = 5;
const DDM_MINUTE_DIGITS = 7;
const GRID_METRE_DIGITS = 3;
const GARS_CELL_DEGREES = 1 / 12;
const GEOREF_PRECISION = 4;

function pointSource(point) {
  return { ...point, sourceKind: "point" };
}

export function fromDecimalDegrees(latitude, longitude) {
  return pointSource(pointFromValues(latitude, longitude));
}

export function fromDms(latitude, longitude) {
  return pointSource(parseCoordinatePair(latitude, longitude));
}

export function fromDdm(latitude, longitude) {
  return pointSource(parseCoordinatePair(latitude, longitude));
}

export function fromMgrs(value) {
  const decoded = decodeMgrs(value, true);
  return {
    latitude: decoded.latitude,
    longitude: decoded.longitude,
    sourceKind: "cell",
    sourceCellMetres: decoded.cellMetres,
  };
}

export function fromUtmUps(zoneValue, hemisphereValue, eastingValue, northingValue) {
  const zoneSource = String(zoneValue).trim();
  if (!/^\d{1,2}$/.test(zoneSource)) {
    throw new Error("Zone must be an integer from 0 to 60.");
  }
  const zone = Number(zoneSource);
  if (zone < 0 || zone > 60) {
    throw new Error("Zone must be an integer from 0 to 60.");
  }
  const hemisphere = String(hemisphereValue).trim().toUpperCase();
  if (hemisphere !== "N" && hemisphere !== "S") {
    throw new Error("Hemisphere must be N or S.");
  }
  const point = utmUpsInverse(
    zone,
    hemisphere === "N",
    parseCoordinateNumber(eastingValue, "Easting"),
    parseCoordinateNumber(northingValue, "Northing"),
  );
  return pointSource(point);
}

export function fromGars(value) {
  const decoded = decodeGars(value);
  return {
    latitude: decoded.latitude,
    longitude: decoded.longitude,
    sourceKind: "area",
    sourceCellDegrees: decoded.cellDegrees,
  };
}

export function fromGeoref(value) {
  const decoded = decodeGeoref(value);
  return {
    latitude: decoded.latitude,
    longitude: decoded.longitude,
    sourceKind: "cell",
    sourceCellDegrees: decoded.cellDegrees,
  };
}

export function coordinateResults(point, mgrsPrecision = 5) {
  const { latitude, longitude } = pointFromValues(
    point.latitude,
    point.longitude,
  );
  const precision = Number(mgrsPrecision);
  if (!Number.isInteger(precision) || precision < 0 || precision > 5) {
    throw new Error("MGRS precision must be between 0 and 5.");
  }

  const grid = utmUpsForward(latitude, longitude);
  const gridPrefix =
    grid.zone === 0
      ? `UPS ${grid.north ? "N" : "S"}`
      : `${grid.zone}${grid.north ? "N" : "S"}`;
  const ddStepDegrees = 10 ** -DD_DIGITS;
  const dmsStepDegrees = 10 ** -DMS_SECOND_DIGITS / 3600;
  const ddmStepDegrees = 10 ** -DDM_MINUTE_DIGITS / 60;
  const mgrsCellMetres = 10 ** (5 - precision);
  const gridStepMetres = 10 ** -GRID_METRE_DIGITS;
  const georefCellDegrees =
    10 ** (2 - GEOREF_PRECISION) / 60;
  return {
    latitude,
    longitude,
    dd: `${fixed(latitude, DD_DIGITS)}, ${fixed(longitude, DD_DIGITS)}`,
    dms: `${formatDms(latitude, "latitude", DMS_SECOND_DIGITS)}  ${formatDms(longitude, "longitude", DMS_SECOND_DIGITS)}`,
    ddm: `${formatDdm(latitude, "latitude", DDM_MINUTE_DIGITS)}  ${formatDdm(longitude, "longitude", DDM_MINUTE_DIGITS)}`,
    mgrs: formatMgrs(encodeMgrs(latitude, longitude, precision)),
    utmUps: `${gridPrefix}  ${fixed(grid.easting, GRID_METRE_DIGITS)} E  ${fixed(grid.northing, GRID_METRE_DIGITS)} N`,
    gars: encodeGars(longitude, latitude),
    georef: encodeGeoref(longitude, latitude, GEOREF_PRECISION),
    resolution: {
      dd: {
        kind: "angular-rounding",
        stepDegrees: ddStepDegrees,
        maximumErrorDegrees: ddStepDegrees / 2,
      },
      dms: {
        kind: "angular-rounding",
        stepDegrees: dmsStepDegrees,
        maximumErrorDegrees: dmsStepDegrees / 2,
      },
      ddm: {
        kind: "angular-rounding",
        stepDegrees: ddmStepDegrees,
        maximumErrorDegrees: ddmStepDegrees / 2,
      },
      mgrs: {
        kind: "grid-cell",
        cellMetres: mgrsCellMetres,
        decodedPoint: "cell-center",
        maximumCenterOffsetMetres:
          mgrsCellMetres * Math.SQRT1_2,
      },
      utmUps: {
        kind: "grid-rounding",
        stepMetres: gridStepMetres,
        maximumErrorMetresPerAxis: gridStepMetres / 2,
      },
      gars: {
        kind: "angular-cell",
        cellDegrees: GARS_CELL_DEGREES,
        decodedPoint: "cell-center",
        maximumCenterOffsetDegreesPerAxis:
          GARS_CELL_DEGREES / 2,
      },
      georef: {
        kind: "angular-cell",
        cellDegrees: georefCellDegrees,
        decodedPoint: "cell-center",
        maximumCenterOffsetDegreesPerAxis:
          georefCellDegrees / 2,
      },
    },
    sourceKind: point.sourceKind ?? "point",
    sourceCellMetres: point.sourceCellMetres,
    sourceCellDegrees: point.sourceCellDegrees,
  };
}

export {
  decodeGars,
  decodeGeoref,
  decodeMgrs,
  encodeGars,
  encodeGeoref,
  encodeMgrs,
  formatDdm,
  formatDms,
  transformCrs,
  utmUpsForward,
  utmUpsInverse,
};
export {
  normalizeEpsg,
  SUPPORTED_CRS,
} from "./projections/crs.js";
